src/v/s3/client.cc

```diff
@@ -775,4 +772,29 @@ void client_pool::release(ss::shared_ptr<client> leased) {
     _cvar.signal();
 }
 
+void client_pool::load_credentials(cloud_roles::credentials&& credentials) {
+    if (unlikely(!_apply_credentials)) {
+        _apply_credentials = ss::make_lw_shared(
+          cloud_roles::make_credentials_applier(std::move(credentials)));
+        init();
+        // We signal the waiter only after the client pool is initialized, so
+        // that any upload operations waiting are ready to proceed.
+        _credentials_var.signal();
+    } else {
+        _apply_credentials->reset_creds(std::move(credentials));
+    }
+}
+
+ss::future<> client_pool::wait_for_credentials() {
+    co_await _credentials_var.wait(max_wait_for_credentials, [this]() {
```

Comment:
I assumed that since we 

1. signal only when `_apply_credentials` is unset, and there we create `_apply_credentials` before signalling
2. wait only when `_apply_credentials` is unset

we would not be able to signal before a wait? 

Basically we make sure that `_apply_credentials` is set before signalling, so that if another fiber is waiting it will get the signal, and if no fiber is waiting, setting `_apply_credentials` will make sure no other fiber will wait, as all of these fibers are going to run on the same core.

One potential issue that I can see is that wait_for_credentials has no invariants saying that `_apply_credentials` need to be unset, maybe I can add a check before the wait such as `if (!_apply_cred)`

is there another way to protect against this, or any other case I am missing?

Reply:
looking at the code, wait with predicate uses `do_until` which first calls the predicate:

```
    template<typename Clock = typename timer<>::clock, typename Duration = typename Clock::duration, typename Pred>
    SEASTAR_CONCEPT( requires seastar::InvokeReturns<Pred, bool> )
    future<> wait(std::chrono::time_point<Clock, Duration> timeout, Pred&& pred) noexcept {
        return do_until(std::forward<Pred>(pred), [this, timeout] {
            return wait(timeout);
        });
    }

...

future<> do_until(StopCondition stop_cond, AsyncAction action) noexcept {
    using namespace internal;
    for (;;) {
      try {
        if (stop_cond()) {
            return make_ready_future<>();
        }
      } catch (...) {
```

Reply:
> I thought that wait calls the predicate before sleep to protect from that.

that's true, but it is a separate issue. I'm only asking about the common pitfalls of using a condition variable which are lost wake ups.